lib/vsc/filesystem/lustre.py

```diff
@@ -0,0 +1,613 @@
+    """ Lustre Operations """
+
+    def __init__(self):
+        super(LustreOperations, self).__init__()
+        self.supportedfilesystems = ['lustre']
+        self.filesystems = {}
+        self.filesets = {}
+
+
+    def _execute_lfs(self, name, opts=None, changes=False):
+        """Return and check the LUSTRE lfs command.
+        """
+
+        cmd = ['/usr/bin/lfs', name]
+
+        if opts is not None:
+            if isinstance(opts, (tuple, list,)):
+                cmd += list(opts)
+            else:
+                self.log.raiseException("_execute_lfs: please use a list or tuple for options: cmd %s opts %s" %
+                                        (cmd, opts), LustreOperationError)
+
+        ec, out = self._execute(cmd, changes)
+
+        return ec, out
+
+    def _execute_lctl_get_param_qmt_yaml(self, device, typ, quotyp='block'):
+        """ executy LUSTRE lctl get_param qmt.* command and parse output
+
+        eg:
+        lctl get_param qmt.kwlust-*.dt*.glb-prj
+        qmt.kwlust-QMT0000.dt-0x0.glb-prj=
+        global_pool0_dt_prj
+        - id:      0
+          limits:  { hard:     0, soft:   0, granted:    0, time:    604800 }
+        - id:      1
+          limits:  { hard:  3798016, soft:   3591168, granted:   3874836, time:  1599748949 }
+        - id:      2
+          limits:  { hard:   0, soft:   0, granted:  0, time:   281474976710656 }
+        - id:      3
+          limits:  { hard:   0, soft:   0, granted:  0, time: 281474976710656 }
+
+        """
+        if typ not in TYP2PARAM:
+            self.log.raiseException("_execute_lctl_get_param_qmt_yaml: unsupported type %s. Use USR,GRP or FILESET"
+                % typ, LustreOperationError)
+        if quotyp not in QUOTYP2PARAM:
+            self.log.raiseException("_execute_lctl_get_param_qmt_yaml: unsupported type %s. Use 'block' or 'inode'" %
+                quotyp, LustreOperationError)
+
+        param = 'qmt.%s-*.%s-*.glb-%s' % (device, QUOTYP2PARAM[quotyp], TYP2PARAM[typ])
+        opts = ['get_param', param]
+        _ec, res = self._execute_lctl(opts)
+        quota_info = res.split("\n", 2)
+        try:
+            newres = yaml.safe_load(quota_info[2])
+        except yaml.YAMLError as exc:
+            self.log.raiseException("_execute_lctl_get_param_qmt_yaml: Error in yaml output: %s"
+                % exc, LustreOperationError)
+
+        return newres
+
+    def _execute_lctl(self, opts, changes=False):
+        """ Return output of lctl command """
+
+        cmd = ['/usr/sbin/lctl']
+        if opts is not None:
+            if isinstance(opts, (tuple, list,)):
+                cmd += list(opts)
+            else:
+                self.log.raiseException("_execute_lctl: please use a list or tuple for options: cmd %s opts %s" %
+                                        (cmd, opts), LustreOperationError)
+
+        ec, res = self._execute(cmd, changes)
+
+        return ec, res
```

Comment:
Merely a suggestion, but should we not check the ec here and raise an exception? Exceptions can always be caught when the caller want to do something about it, but I'm not sure we should have the caller check the ec and raise an exception there. In either case, the caller will be in the trace, and I think I'd prefer it if we have

~~~~
caller line X: _execute_lctl
_execute_lctl: line Y raised the exception
~~~~

rather than
~~~~
caller raised exception on line X+1
~~~~